netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -365,10 +437,22 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        * lifetime and we only want to configure it once.
        */
       if (handlers != null) {
-        ctx.pipeline().addFirst(handlers);
+        for (ChannelHandler handler : handlers) {
+          ctx.pipeline().addBefore(ctx.name(), null, handler);
+        }
+        ChannelHandler handler0 = handlers[0];
+        ChannelHandlerContext handler0Ctx = ctx.pipeline().context(handlers[0]);
         handlers = null;
+        if (handler0Ctx != null) { // The handler may have removed itself immediately
```

Comment:
Does this comment apply to the case where `handler0Ctx == null`, where we don't need to call `channelRegistered()` on it?

To confirm my understanding, take `BufferUntilProxyTunnelledHandler(HttpProxyHandler, BufferUntilTlsNegotiatedHandler` for example, when  `BufferUntilProxyTunnelledHandler.channelRegistered()` is running, the pipe line will become
```
[HttpProxyHandler, BufferUntilTlsNegotiatedHandler, BufferUntilProxyTunnelledHandler]
```
Thus `handler0` will be `HttpProxyHandler`. Calling `handler0.fireChannelRegistered()` will call `fireChannelRegistered()` on every handler in the pipeline, including `BufferUntilTlsNegotiatedHandler` which would turn the pipeline into:
```
[HttpProxyHandler, SslHandler, BufferUntilTlsNegotiatedHandler, BufferUntilProxyTunnelledHandler]
```

Am I right?